### PR TITLE
move metric client to extensions nuget

### PR DESCRIPTION
**Pull request by bermanguy**

### CODE REVIEW

**Comment by aviled** [comment]
>     public static class PolicyExtensions

I don't like that this is public for anyone that uses the package.
This contains really internal stuff.

Please make this internal and move clients that uses this to the extension packages 

If you can't then consider not using this class in the metric client.

---
Refers to: src/SDK/SmartDetectors.Extensions/PolicyExtensions.cs:19 in 098b2ec. [](commit_id = 098b2ec775439e0b77f3be3263e0eba75ef55bc3, deletion_comment = False)

**Comment by aviled** [comment]
>   <ItemGroup>

you need to add package dependencies here, e.g. Microsoft.Azure.Management.Monitor.Fluent

---
Refers to: src/SDK/SmartDetectorsExtensionsPackage/SmartDetectorsExtensionsPackage.nuproj:45 in 098b2ec. [](commit_id = 098b2ec775439e0b77f3be3263e0eba75ef55bc3, deletion_comment = False)

**Comment by aviled** [comment]
>     public static class PolicyExtensions

Let's discuss this and decide if it is worth the effort.

---
In reply to: [463080805](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/48#issuecomment-463080805) [](ancestors = 463080805)

---
Refers to: src/SDK/SmartDetectorsShared/Extensions/PolicyExtensions.cs:19 in 098b2ec. [](commit_id = 098b2ec775439e0b77f3be3263e0eba75ef55bc3, deletion_comment = False)

**Comment by liadtal** [comment]
>     public static class PolicyExtensions

Agree with Aviram. Let's keep the IExtendedTracer and the PolicyExtensions outside of this Nuget.
Since the MetricClient only needs to call RunAndTrackDependencyAsync, why don't we add to its ctor an optional interface that implements just this method?

---
In reply to: [463081291](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/48#issuecomment-463081291) [](ancestors = 463081291,463080805)

---
Refers to: src/SDK/SmartDetectorsShared/Extensions/PolicyExtensions.cs:19 in 098b2ec. [](commit_id = 098b2ec775439e0b77f3be3263e0eba75ef55bc3, deletion_comment = False)

**Comment by bermanguy** [comment]
>     public static class PolicyExtensions

moved IcredentialsFactory & PolicyExtensions back to Shared code, reduced the metric client in the Extensions nuget to include only the logic. Created an extended client in the appliance Shared code that wrap the Metric Client and track & trace  & uses retry policy

---
In reply to: [463096636](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/48#issuecomment-463096636) [](ancestors = 463096636,463081291,463080805)

---
Refers to: src/SDK/SmartDetectorsShared/Extensions/PolicyExtensions.cs:19 in 098b2ec. [](commit_id = 098b2ec775439e0b77f3be3263e0eba75ef55bc3, deletion_comment = False)

File: src/SDK/SmartDetectorsShared/SmartDetectorsShared.csproj [hunk 1]
```diff
@@ -20,7 +20,6 @@
     <Reference Include="System.Xml" />
   </ItemGroup>
   <ItemGroup>
-    <PackageReference Include="Microsoft.Azure.Management.Monitor.Fluent" Version="1.11.1" />
```
Comment (aviled) [inline src/SDK/SmartDetectorsShared/SmartDetectorsShared.csproj]: >Monitor [](start = 58, length = 7)

you need this to create the monitor management client

File: src/SDK/SmartDetectorsShared/Clients/ExtendedMetricClient.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        /// Get the resource metric definitions, based on the resource and service (for example: if the resource is a storage account, possible services are BLOB, Queue, Table and File)
+        /// </summary>
+        /// <param name="resource">The Azure resource for which we want to fetch metric definitions</param>
+        /// <param name="azureResourceService">The Azure resource's service type</param>
+        /// <param name="cancellationToken">Cancellation Token for the async operation</param>
+        /// <returns>A <see cref="Task{TResult}"/> object that represents the asynchronous operation, returning the list of metric definitions</returns>
+        public async Task<IEnumerable<MetricDefinition>> GetResourceMetricDefinitionsAsync(ResourceIdentifier resource, ServiceType azureResourceService, CancellationToken cancellationToken)
+        {
+            string resourceFullUri = MetricClient.GetResourceFullUri(resource, azureResourceService);
+            return await this.GetResourceMetricDefinitionsAsync(resourceFullUri, cancellationToken);
```
Comment (liadtal) [inline src/SDK/SmartDetectorsShared/Clients/ExtendedMetricClient.cs]: Call metricClient.GetResourceMetricDefinitionsAsync(resource, ..)
No need to call GetResourceFullUri here

File: src/SDK/SmartDetectorsShared/Clients/ExtendedMetricClient.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        /// </summary>
+        /// <param name="resource">The Azure resource for which we want to fetch metrics</param>
+        /// <param name="azureResourceService">The Azure resource's service type</param>
+        /// <param name="queryParameters">Query parameters to be used when fetching metric data. All fields are optional</param>
+        /// <param name="cancellationToken">Cancellation Token for the async operation</param>
+        /// <returns>A <see cref="Task{TResult}"/> object that represents the asynchronous operation, returning the list metrics</returns>
+        public async Task<IEnumerable<MetricQueryResult>> GetResourceMetricsAsync(ResourceIdentifier resource, ServiceType azureResourceService, QueryParameters queryParameters, CancellationToken cancellationToken)
+        {
+            string resourceFullUri = MetricClient.GetResourceFullUri(resource, azureResourceService);
+            return await this.GetResourceMetricsAsync(resourceFullUri, queryParameters, cancellationToken);
```
Comment (liadtal) [inline src/SDK/SmartDetectorsShared/Clients/ExtendedMetricClient.cs]: Same - call metricClient.GetResourceMetricsAsync(resource, ...)

File: src/SDK/SmartDetectors.Extensions/Clients/MetricClient.cs [hunk 4]
```diff
@@ -40,41 +36,38 @@ public class MetricClient : IMetricClient
...
-        /// <param name="tracer">The tracer</param>
-        /// <param name="credentialsFactory">The credentials factory</param>
-        /// <param name="subscriptionId">The subscription Id</param>
-        public MetricClient(IExtendedTracer tracer, ICredentialsFactory credentialsFactory, string subscriptionId)
-            : this(tracer, subscriptionId, new MonitorManagementClient(credentialsFactory.Create("https://management.azure.com/")))
+        /// <param name="resource">The Azure resource for which we want to fetch metrics</param>
+        /// <param name="azureResourceService">The Azure resource's service type</param>
+        /// <returns>The full Resource metrics Uri</returns>
+        [SuppressMessage("Microsoft.Design", "CA1055:UriReturnValuesShouldNotBeStrings", Justification = "Keeping alignment with the Microsoft.Azure.Management.Monitor.Fluent.MetricDefinitionsOperationsExtensions API")]
+        public static string GetResourceFullUri(ResourceIdentifier resource, ServiceType azureResourceService)
```
Comment (liadtal) [inline src/SDK/SmartDetectors.Extensions/Clients/MetricClient.cs]: >GetResourceFullUri [](start = 29, length = 18)

This shouldn't be public

File: src/SDK/SmartDetectorsShared/Clients/ExtendedMetricClient.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    using Microsoft.Azure.Management.Monitor.Fluent;
+    using Microsoft.Azure.Monitoring.SmartDetectors.Extensions;
+    using Microsoft.Azure.Monitoring.SmartDetectors.Extensions.Clients;
+    using Microsoft.Azure.Monitoring.SmartDetectors.Metric;
+    using Microsoft.Azure.Monitoring.SmartDetectors.Tools;
+    using Microsoft.Azure.Monitoring.SmartDetectors.Trace;
+    using Polly;
+
+    /// <summary>
+    /// An extended metric client that implements <see cref="IMetricClient"/> that wrapped by retry policy and tracks the results and.
```
Comment (liadtal) [inline src/SDK/SmartDetectorsShared/Clients/ExtendedMetricClient.cs]: >that wrapped by retry policy and tracks the results and [](start = 78, length = 55)

That wraps an existing IMetricClient and adds traces and dependency tracking

File: src/SDK/SmartDetectorsExtensionsPackage/SmartDetectorsExtensionsPackage.nuproj [hunk 6]
```diff
@@ -22,7 +22,7 @@
   <Import Project="$(NuProjPath)\NuProj.props" Condition="Exists('$(NuProjPath)\NuProj.props')" />
   <PropertyGroup Label="Configuration">
     <Id>Microsoft.Azure.Monitoring.SmartDetectors.Extensions</Id>
-    <Version>0.16.0.0</Version>
+    <Version>0.17.0.0</Version>
```
Comment (liadtal) [inline src/SDK/SmartDetectorsExtensionsPackage/SmartDetectorsExtensionsPackage.nuproj]: >17 [](start = 15, length = 2)

There was no breaking or major change in the SmartDetectorsSDK. Consider just updating the minor version, to 0.16.1.0

File: src/SDK/MonitoringApplianceEmulator/Models/AuthenticationServices.cs [hunk 7]
```diff
@@ -106,7 +106,8 @@ public async Task<string> GetResourceTokenAsync(string resource)
                     // Check again
                     if (this.IsAccessTokenAboutToExpire)
                     {
-                        this.authenticationResult = this.authenticationContext.AcquireTokenByRefreshToken(this.authenticationResult.RefreshToken, this.clientId);
+                        // This should be fixed. opened bug for tracking #3942846
```
Comment (liadtal) [inline src/SDK/MonitoringApplianceEmulator/Models/AuthenticationServices.cs]: >This [](start = 27, length = 4)

Add "TODO:" at the beginning, this way it's appear as a resharper warning

File: src/SDK/MonitoringApplianceEmulator/Models/AuthenticationServices.cs [hunk 8]
```diff
@@ -80,13 +80,13 @@ public AuthenticationServices()
         /// </summary>
         public void AuthenticateUser()
         {
```
Comment (aviled) [inline src/SDK/MonitoringApplianceEmulator/Models/AuthenticationServices.cs]: Can you make this async Task AuthenticateUserAsync() and use await inside?

File: src/SDK/SmartDetectors.Extensions/ResourceIdentifierExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+    /// <summary>
+    /// Extension methods for <see cref="ResourceIdentifier"/>
+    /// </summary>
+    public static class ResourceIdentifierExtensions
+    {
+        /// <summary>
+        /// A dictionary, mapping <see cref="ServiceType"/> enumeration values to matching presentation in URI
+        /// </summary>
+        public static readonly ReadOnlyDictionary<ServiceType, string> MapAzureServiceTypeToPresentationInUri =
```
Comment (aviled) [inline src/SDK/SmartDetectors.Extensions/ResourceIdentifierExtensions.cs]: >public [](start = 8, length = 6)

can this be private?

File: src/SDK/SmartDetectorsExtensionsPackage/SmartDetectorsExtensionsPackage.nuproj [hunk 10]
```diff
@@ -43,6 +43,9 @@
     <ProjectReference Include="..\SmartDetectorsPackage\SmartDetectorsPackage.nuproj" />
   </ItemGroup>
   <ItemGroup>
+    <Dependency Include="Microsoft.Azure.Management.Monitor.Fluent">
+      <Version>1.11.1</Version>
```
Comment (aviled) [inline src/SDK/SmartDetectorsExtensionsPackage/SmartDetectorsExtensionsPackage.nuproj]: >1.11.1 [](start = 15, length = 6)

This is not the correct version

File: src/SDK/SmartDetectorsShared/Clients/MetricClient.cs [hunk 11]
```diff
@@ -136,13 +130,13 @@ public async Task<IEnumerable<MetricQueryResult>> GetResourceMetricsAsync(string
                     metricnames: queryParameters.MetricNames == null ? string.Empty : string.Join(",", queryParameters.MetricNames),
                     aggregation: queryParameters.Aggregations != null ? string.Join(",", queryParameters.Aggregations) : null,
                     top: queryParameters.Top,
-                    orderby: queryParameters.Orderby,
+                    @orderby: queryParameters.Orderby,
```
Comment (aviled) [inline src/SDK/SmartDetectorsShared/Clients/MetricClient.cs]: >@ [](start = 20, length = 1)

what is this?

File: src/SDK/MonitoringApplianceEmulator/Models/ActiveDirectoryCredentialsFactory.cs [hunk 12]
```diff
@@ -30,9 +32,25 @@ public ActiveDirectoryCredentialsFactory(IAuthenticationServices authenticationS
...
         }
+
+        /// <summary>
+        /// Create an instance of the <see cref="AzureCredentials"/> class.
+        /// </summary>
+        /// <param name="resource">The resource for which to create the credentials</param>
+        /// <returns>The credentials</returns>
+        public AzureCredentials CreateAzureCredentials(string resource)
+        {
+            var activeDirectoryCredentials = new ActiveDirectoryCredentials(this.authenticationServices, resource);
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/Models/ActiveDirectoryCredentialsFactory.cs]: >new ActiveDirectoryCredentials(this.authenticationServices, resource); [](start = 45, length = 70)

this.CreateServiceClientCredentials

File: src/SDK/MonitoringApplianceEmulator/Models/AuthenticationServices.cs [hunk 13]
```diff
@@ -106,7 +106,8 @@ public async Task<string> GetResourceTokenAsync(string resource)
                     // Check again
                     if (this.IsAccessTokenAboutToExpire)
                     {
-                        this.authenticationResult = this.authenticationContext.AcquireTokenByRefreshToken(this.authenticationResult.RefreshToken, this.clientId);
+                        // This should be fixed. opened bug for tracking #3942846
+                        //// this.authenticationResult = this.authenticationContext.AcquireTokenByRefreshToken(this.authenticationResult.RefreshToken, this.clientId);
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/Models/AuthenticationServices.cs]: >//// this.authenticationResult = this.authenticationContext.AcquireTokenByRefreshToken(this.authenticationResult.RefreshToken, this.clientId); [](start = 24, length = 142)

why?

File: src/SDK/SmartDetectors.Extensions/MetricDefinitionExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    /// Extension methods for <see cref="Management.Monitor.Fluent.Models.MetricDefinition"/>
+    /// </summary>
+    public static class MetricDefinitionExtensions
+    {
+        /// <summary>
+        /// Converts a <see cref="Management.Monitor.Fluent.Models.MetricDefinition"/> response to a <see cref="MetricDefinition"/> and returns it.
+        /// </summary>
+        /// <param name="definition">The metric definition</param>
+        /// <returns>The conversion result</returns>
+        public static MetricDefinition ConvertToSmartDetectorsMetricDefinition(this Management.Monitor.Fluent.Models.MetricDefinition definition)
```
Comment (EranGon) [inline src/SDK/SmartDetectors.Extensions/MetricDefinitionExtensions.cs]: >ConvertToSmartDetectorsMetricDefinition [](start = 39, length = 39)

ToSmartDetectorsMetricDefinition

File: src/SDK/SmartDetectors.Extensions/ResponseInnerExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    /// Extension methods for <see cref="ResponseInner"/>
+    /// </summary>
+    public static class ResponseInnerExtensions
+    {
+        /// <summary>
+        /// Converts a <see cref="ResponseInner"/> to a list of <see cref="MetricQueryResult"/> and returns it
+        /// </summary>
+        /// <param name="queryResponse">The metric query response as returned by Azure Monitoring</param>
+        /// <returns>A list of metric query results</returns>
+        public static IList<MetricQueryResult> ConvertToSmartDetectorsMetricQueryResult(this ResponseInner queryResponse)
```
Comment (EranGon) [inline src/SDK/SmartDetectors.Extensions/ResponseInnerExtensions.cs]: >ConvertToSmartDetectorsMetricQueryResult [](start = 47, length = 40)

ToMetricQueryResult

File: src/SDK/SmartDetectorsShared/Trace/ConsoleTracer.cs [hunk 16]
```diff
@@ -8,6 +8,7 @@ namespace Microsoft.Azure.Monitoring.SmartDetectors.Trace
 {
     using System;
     using System.Collections.Generic;
+    using Microsoft.Azure.Monitoring.SmartDetectors.Extensions;
```
Comment (EranGon) [inline src/SDK/SmartDetectorsShared/Trace/ConsoleTracer.cs]: >Microsoft.Azure.Monitoring.SmartDetectors.Extensions [](start = 10, length = 52)

why?

File: src/SmartDetectorsApplianceShared/MsiCredentialsFactory.cs [hunk 17]
```diff
@@ -34,9 +36,25 @@ public MsiCredentialsFactory(IHttpClientWrapper httpClientWrapper, IExtendedTrac
...
         }
+
+        /// <summary>
+        /// Create an instance of the <see cref="AzureCredentials"/> class.
+        /// </summary>
+        /// <param name="resource">The resource for which to create the credentials</param>
+        /// <returns>The credentials</returns>
+        public AzureCredentials CreateAzureCredentials(string resource)
+        {
+            var msiCredentials = new MsiCredentials(this.httpClientWrapper, resource, this.tracer);
```
Comment (EranGon) [inline src/SmartDetectorsApplianceShared/MsiCredentialsFactory.cs]: >ew MsiCredentials(this.httpClientWrapper, resource, this.tracer); [](start = 34, length = 65)

this.ServiceClientCredentials

File: src/SmartDetectorsApplianceShared/Trace/TracerFactory.cs [hunk 18]
```diff
@@ -10,6 +10,7 @@ namespace Microsoft.Azure.Monitoring.SmartDetectors.MonitoringAppliance.Trace
     using System.Collections.Generic;
     using Microsoft.ApplicationInsights.Extensibility;
     using Microsoft.ApplicationInsights.WindowsServer.TelemetryChannel;
+    using Microsoft.Azure.Monitoring.SmartDetectors.Extensions;
```
Comment (EranGon) [inline src/SmartDetectorsApplianceShared/Trace/TracerFactory.cs]: >Microsoft.Azure.Monitoring.SmartDetectors.Extensions [](start = 10, length = 52)

remove

File: test/SmartDetectorsApplianceSharedTests/BlobStateRepositoryTests.cs [hunk 19]
```diff
@@ -12,7 +12,7 @@ namespace SmartDetectorsApplianceSharedTests
     using System.Net;
     using System.Threading;
     using System.Threading.Tasks;
-    using Microsoft.Azure.Monitoring.SmartDetectors;
+    using Microsoft.Azure.Monitoring.SmartDetectors.Extensions;
```
Comment (EranGon) [inline test/SmartDetectorsApplianceSharedTests/BlobStateRepositoryTests.cs]: >Extensions [](start = 52, length = 10)

remove

File: test/SmartDetectorsApplianceSharedTests/SmartDetectorRepositoryTest.cs [hunk 20]
```diff
@@ -14,6 +14,7 @@ namespace SmartDetectorsApplianceSharedTests
     using System.Threading;
     using System.Threading.Tasks;
     using Microsoft.Azure.Monitoring.SmartDetectors;
+    using Microsoft.Azure.Monitoring.SmartDetectors.Extensions;
```
Comment (EranGon) [inline test/SmartDetectorsApplianceSharedTests/SmartDetectorRepositoryTest.cs]: >Extensions [](start = 52, length = 10)

remove

File: test/SmartDetectorsExtensionsTests/SmartDetectorsExtensionsTests.csproj [hunk 21]
```diff
@@ -16,6 +16,7 @@
     <PackageReference Include="StyleCop.Analyzers" Version="1.0.2" />
   </ItemGroup>
   <ItemGroup>
+    <ProjectReference Include="..\..\src\SDK\MonitoringApplianceEmulator\MonitoringApplianceEmulator.csproj" />
```
Comment (EranGon) [inline test/SmartDetectorsExtensionsTests/SmartDetectorsExtensionsTests.csproj]: >MonitoringApplianceEmulator [](start = 73, length = 27)

why?

File: src/SDK/SmartDetectorsShared/Trace/ConsoleTracer.cs (hunk 16 above)
Comment (bermanguy) [reply src/SDK/SmartDetectorsShared/Trace/ConsoleTracer.cs]: it was as a result of 100 refactoring I did in this PR.. removed all unnecessary ones

---
In reply to: [257936977](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/48#discussion_r257936977) [](ancestors = 257936977)

File: test/SmartDetectorsExtensionsTests/SmartDetectorsExtensionsTests.csproj (hunk 21 above)
Comment (bermanguy) [reply test/SmartDetectorsExtensionsTests/SmartDetectorsExtensionsTests.csproj]: removed

---
In reply to: [257937497](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/48#discussion_r257937497) [](ancestors = 257937497)

**Pull request merged**